packages/vm/src/compatibility.rs

```diff
@@ -101,62 +99,82 @@ const MAX_TOTAL_FUNCTION_PARAMS: usize = 10_000;
 /// during static validation.
 const MAX_FUNCTION_RESULTS: usize = 1;
 
-#[derive(Clone)]
+#[derive(Clone, Default)]
 pub enum Logs {
     On(RefCell<Vec<String>>),
+    #[default]
     Off,
 }
 
+pub struct LogAccess<'a>(RefMut<'a, Vec<String>>);
+
+impl<'a> LogAccess<'a> {
+    /// Adds a message to the logs
+    pub fn add(&mut self, msg: impl Into<String>) {
+        self.0.push(msg.into());
+    }
+}
+
 impl Logs {
     pub fn new() -> Self {
         On(RefCell::new(Vec::new()))
     }
 
-    // Gets access to logs for writing
-    pub fn open(&mut self) -> Option<&mut Vec<String>> {
+    // Gets access to logs for writing.
+    // Use this if you want to add multiple messages at once.
+    pub fn open(&mut self) -> Option<LogAccess> {
         match self {
-            On(data) => {
-                let mut data = data.borrow_mut();
-                Some(data.as_mut())
-            }
+            On(data) => Some(LogAccess(data.borrow_mut())),
             Off => None,
         }
     }
 
-    pub fn iter(&self) -> impl Iterator<Item = String> {
-        let iter = match self {
-            On(data) => data.borrow().iter(),
-            Off => Vec::new().into_iter().into(), // How to create am empty Iter<String> ?!?!?!?
-        };
-        iter
+    /// Adds a message to the logs, if they are enabled.
+    /// This is a convenience method for adding a single message.
+    ///
+    /// Takes a closure that returns the message to add to avoid unnecessary allocations.
+    pub fn add(&mut self, msg_fn: impl FnOnce() -> String) {
+        if let Some(mut logs) = self.open() {
+            logs.add(msg_fn());
+        }
+    }
+}
+
+impl IntoIterator for Logs {
+    type Item = String;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        match self {
+            Logs::On(data) => data.into_inner().into_iter(),
+            Logs::Off => vec![].into_iter(),
+        }
     }
 }
 
 use Logs::*;
 
 /// Checks if the data is valid wasm and compatibility with the CosmWasm API (imports and exports)
 pub fn check_wasm(wasm_code: &[u8], available_capabilities: &HashSet<String>) -> VmResult<()> {
-    check_wasm_with_logs(wasm_code, available_capabilities, Off)
+    check_wasm_with_logs(wasm_code, available_capabilities, &mut Off)
```

Comment:
Hmm, this is what I wanted to avoid somehow in order to keep the default caller code simple, especially for the tests. Is there a way to use owned `Logs` that can be cloned and use internal mutability?

Reply:
That should be possible with an `Rc<RefCell<...>>` with some runtime overhead.
As an alternative, we could also just print immediately instead of storing in a Vec. Then we don't need a mutable reference, so it could become `&Off` or just `Off` if we use an `Rc` for the `On` case.

Reply:
My thinking was that the CLI should be able to do thinks like
- Decide between output (STDOUT/STDERR/...)
- Add indentation or prefix text to better structure output for multiple contracts

> That should be possible with an Rc<RefCell<...>> with some runtime overhead.

That sounds cool. As logging is not enabled in production anyways, this overhead does not matter

Reply:
Oh, I see, but that could also be achieved by making the whole thing a `Logger` trait and then the impl of that can choose whatever prefix or indentation it wants to add, as well as what output.

Reply:
Or just config in the `Logs::On { ...` case

Reply:
True story. There is no real point in bringing the strings back to the caller code as a vector.

Reply:
Added a config to the `Logs::On` case. Sadly that makes it now impossible to unit test the logger, since it logs directly to stdout / stderr, but in exchange for that it now implements `Copy` and you can just directly pass it in without taking any references or even having to clone.